src/redisai.c

```diff
@@ -1125,129 +1212,154 @@ int RedisAI_ModelList_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv
+        } else {
+          const int get_result = RAI_getTensorFromLocalContext(
+              ctx, *localContextDict, arg_string, &inputTensor);
+          if (get_result == REDISMODULE_ERR) {
+            return -1;
+          }
         }
-        RedisModule_CloseKey(tensorKey);
+
         // Opname here is passed without copying
         const char *opname = NULL;
-        if (mto->inputs) {
-          opname = mto->inputs[ninputs];
+        if ((*mto)->inputs) {
+          opname = (*mto)->inputs[ninputs];
         }
-        if (!RAI_ModelRunCtxAddInput(rinfo->mctx, 0, opname, inputTensor)) {
+        if (!RAI_ModelRunCtxAddInput((*rinfo)->mctx, 0, opname, inputTensor)) {
           // todo free rinfo
-          return RedisModule_ReplyWithError(ctx, "ERR Input key not found");
+          RedisModule_ReplyWithError(ctx, "ERR Input key not found");
+          return -1;
         }
         ninputs++;
       } else {
         // Opname here is passed without copying
         const char *opname = NULL;
-        if (mto->outputs) {
-          opname = mto->outputs[noutputs];
+        if ((*mto)->outputs) {
+          opname = (*mto)->outputs[noutputs];
         }
-        if (!RAI_ModelRunCtxAddOutput(rinfo->mctx, 0, opname)) {
+        if (!RAI_ModelRunCtxAddOutput((*rinfo)->mctx, 0, opname)) {
           // todo free rinfo
           return RedisModule_ReplyWithError(ctx, "ERR Output key not found");
         }
-        rinfo->outkeys[noutputs] = argv[argpos];
+        (*rinfo)->outkeys[noutputs] = argv[argpos];
         noutputs++;
       }
     }
   }
 
-  if (mto->inputs && array_len(mto->inputs) != ninputs) {
+  if ((*mto)->inputs && array_len((*mto)->inputs) != ninputs) {
     return RedisModule_ReplyWithError(
         ctx,
-        "Number of names given as INPUTS during MODELSET and keys given as INPUTS here do not match");
+        "Number of names given as INPUTS during MODELSET and keys given as "
+        "INPUTS here do not match");
   }
 
-  if (mto->outputs && array_len(mto->outputs) != noutputs) {
+  if ((*mto)->outputs && array_len((*mto)->outputs) != noutputs) {
     return RedisModule_ReplyWithError(
         ctx,
-        "Number of names given as OUTPUTS during MODELSET and keys given as OUTPUTS here do not match");
+        "Number of names given as OUTPUTS during MODELSET and keys given as "
+        "OUTPUTS here do not match");
+  }
+  return argpos;
+}
+
+/**
+ * AI.MODELRUN model_key INPUTS input_key1 ... OUTPUTS output_key1 ...
+ *
+ * The request is queued and evaded asynchronously from a separate thread. The
+ * client blocks until the computation finishes.
+ *
+ * 1. clone inputs as needed in the main thread (only the alternative is to
+ * lock)
+ * 2. spawn the new thread for running the model
+ * 3. have reply callback put the data back into the key
+ * 
+ * This way we avoid any race condition. The only gotcha is making sure no one
+ * overwrites the model until it's done computing.
```

Comment:
I find this comment a bit misleading:
`RAI_ModelRunCtxCreate` will shallow copy the model, so since this command runs in the main thread, there's no need for locking or waiting because run context will hold the model that was found in the key at the time this portion of the command was executed. If a new model is set while the run is waiting in the queue, the model is disposed at the key level, but the old model is retained in the run context and runs. It is then deallocated when run context is deallocated.

Reply:
@lantiga agree. This was an old comment that just moved from inside the function to the function description. 
https://github.com/RedisAI/RedisAI/blob/master/src/redisai.c#L1139
Should we PR this change specifically on a different pull request, merge it on master and then update here? Or just updating here is enough?

Reply:
Right, let’s just update here. I’d like to fast track this work anyway

Reply:
> Looking very good so far.
> 
> I propose to refactor RunInfo, so that instead of a single mctx and sctx pointer, it holds an array of (mctx | sctx) pointers. I think it's better to do it now than wait too long and have to change code pervasively again. This will open the door to the full DAG implementation which at this point I'd like to get done shortly. 
> 

@lantiga the latest commit makes usage of this change for DAGRUN, so that instead of a single mctx and sctx pointer, it holds an array of (mctx | sctx) pointers (within RAI_DagOp). To be gradually adopted on modelrun and scriptrun ( for now and to be able to fully test and harden it only on dagrun ). 
What do you think?